[sharktank] Add mixture of experts(moe) support for Llama 4 model

vivekkhandelwal1 opened the pull request:
This commit adds support for mixture of experts for the Llama 4 model.

The code is adopted from https://github.com/nod-ai/shark-ai/tree/users/sogartar/llama4.

Authored-by: Boian Petkantchin <Boian.Petkantchin@amd.com>
Co-authore-by: Vivek Khandelwal <vivekkhandelwal1424@gmail.com>

sogartar commented on sharktank/sharktank/layers/ffn_moe_block.py:
```diff
@@ -32,21 +32,47 @@ def __init__(
         inputs = inputs[:, :]
         weights = weights[experts, :, :]
-        matmul = einsum_2args(inputs, weights, "mek,menk->men")
+        if expert_gate is not None:
+            weights = weights * expert_gate.view(
+                expert_gate.shape[0], expert_gate.shape[1], 1, 1
+            )
```
Actually, this is concerning as I think we would not be hitting this path as I look the rest of the code and also on the [users/sogartar/llama4](https://github.com/nod-ai/shark-ai/tree/users/sogartar/llama4) branch.

sogartar replied:
Disregard what I wrote earlier about this block. We are actually multiplying by the `expert_gate` but after this operation. It is OK to do that as these operations commute.
This is all throw-away code. The only thin that is really different in this class is that we don't post multiply with the `expert_gate`/`router_scores` as mentioned in the other comment.

vivekkhandelwal1 replied:
Hi Boian,sorry but I am confused by your comment. Can you please clarify whether we need this or not?

sogartar replied:
I meant that you can remove this bit + the `expert_gate` argument as it will be unused.

sogartar commented on sharktank/sharktank/layers/configs/llm_configs.py:
```diff
@@ -221,6 +272,55 @@ class LlamaModelConfig:
+    # If True, the Feed-Forward Network (FFN) block adds the residual connection at the end.
+    ffn_add_residual: bool = True
+
+    # The size of the model's vocabulary.
+    vocabulary_size: Optional[int] = None
```
I was wrong putting this here. It should be in `LlamaHParams` as it is available in GGUF. there also would need to be addition in the `from_gguf_prop` and `to_gguf_props` methods.
The difference is that there it is named `vocab_size`.

sogartar commented on sharktank/sharktank/layers/ffn_moe_block.py:
```diff
@@ -32,21 +32,47 @@ def __init__(
         self.activation_fn = activation_fn
 
-    def pre_matmul_gather(self, inputs, weights, experts, einstring="mk,menk->men"):
+    def pre_matmul_gather_orig(
+        self,
+        inputs: torch.Tensor,
+        weights: torch.Tensor,
+        # (batch_size * sequence_length, num_top_experts)
+        experts: torch.Tensor,
+        einstring="mk,menk->men",
+    ):
```
This can be removed.

sogartar commented on sharktank/sharktank/layers/ffn_moe_block.py:
```diff
@@ -32,21 +32,47 @@ def __init__(
+        # (batch_size * sequence_length, num_top_experts)
+        experts: torch.Tensor,
+        # (bs * sl, num_top_experts)
+        expert_gate: torch.Tensor | None = None,
```
This argument should be removed.

sogartar commented on sharktank/sharktank/models/llama/testing.py:
```diff
@@ -101,54 +104,77 @@ def make_attention_block_ffn_theta_v2(
 
-def make_moe_block_theta(feature_dim=1024, ffn_dim=6144, num_experts=8) -> Theta:
-    return Theta(
+def make_attention_moe_block_random_theta(
```
We must make sure to reconcile with the work coming from the Deepseek side. There are probably discrepancies.

archana-ramalingam replied:
Deepseek, grok have their own `make_moe_theta()`. This change shouldn't affect them. We need to refactor all these functions into a single `testing.py`.

archana-ramalingam commented on sharktank/sharktank/layers/configs/llm_configs.py:
```diff
@@ -150,8 +188,12 @@ def _optional_float_prop(p: dict[str, Any], name: str, default_value: float) ->
-def _optional_int_prop(p: dict[str, Any], name: str, default_value: int) -> int:
+def _optional_int_prop(
+    p: dict[str, Any], name: str, default_value: Optional[int] = None
+) -> int:
```
If we are returning int and None, shouldn't this be `-> Union[int, None]:` ?

vivekkhandelwal1 replied:
It seems you have added this change.

archana-ramalingam commented on sharktank/sharktank/layers/ffn_block.py:
```diff
@@ -60,4 +62,7 @@ def forward(
             ffn_activation = ops.elementwise(self.activation_fn, ffn_up)
             ffn_down = self.ffn_down(ffn_activation)
 
+        if self.add_residual:
```
Determine the add_residual version for llama4 and move accordingly.

vivekkhandelwal1 replied:
Hi Archana, I took a look at the Llama4 code and the residual is added after the sum up of shared and routed experts in MOE. You can see [here](https://github.com/huggingface/transformers/blob/b369a65480cf1df22b3c853f086b832bc5785a19/src/transformers/models/llama4/modeling_llama4.py#L448) that the addition of residual happens after the execution of MOE block [here](https://github.com/huggingface/transformers/blob/b369a65480cf1df22b3c853f086b832bc5785a19/src/transformers/models/llama4/modeling_llama4.py#L443). Therefore, we have to keep this change.

archana-ramalingam replied:
So we will have to remove all `add_residual` in `ffn_block.py` because this op has to happen at the end of ffn block in llm.py.

archana-ramalingam commented on sharktank/sharktank/layers/mixture_of_experts_block.py:
```diff
@@ -110,7 +110,7 @@ def __init__(
             )
             self.shared_experts = FFN(
-                theta=shared_ffn_theta, activation_fn=moe_activation
+                theta=shared_ffn_theta, activation_fn=moe_activation, add_residual=False
```
Same comment as above for `add_residual`

archana-ramalingam commented on sharktank/sharktank/models/llm/llm.py:
```diff
@@ -377,14 +383,25 @@ def __init__(
+                moe_activation,
+                self.add_residual,
+                normalize_experts,
+            ) = moe_func_map[config.hp.model_arch]
```
Can move this outside the `if is_moe_block:`, so that `self.add_residual` is visible later on.

archana-ramalingam commented on sharktank/sharktank/models/llm/llm.py:
```diff
@@ -407,6 +424,8 @@ def __init__(
                     theta=theta,
                     fake_quant=fake_quant,
+                    activation_fn=activation_fn,
+                    add_residual=config.ffn_add_residual,
```
Same comment for `add_residual`. Ideally we want all models to fetch `add_residual` from config, till then self.add_residual will be `config.ffn_add_residual` or `self.add_residual` in this script.

vivekkhandelwal1 replied:
I have made the change. Please let me know if it's correct or not.

archana-ramalingam replied:
Shouldn't have added my comment here. add_residual has to go anyway. Refer to comment below.

vivekkhandelwal1 commented on sharktank/sharktank/models/llama/testing.py:
```diff
@@ -101,54 +104,77 @@ def make_attention_block_ffn_theta_v2(
-                name="blk.0.ffn_down_exps.weight",
-                data=make_rand_torch((num_experts, ffn_dim, feature_dim * num_experts)),
-            ),
+            "ffn_norm.weight": DefaultPrimitiveTensor(
```
Fixed now.

archana-ramalingam commented on sharktank/sharktank/layers/configs/llm_configs.py:
```diff
@@ -81,17 +90,17 @@ class LlamaHParams:
     n_dense_layers: Optional[int] = None
     route_scale: Optional[float] = None
 
-    # Deepseek MoE config
-    expert_shared_count: Optional[int] = None
```
I see this change was missed while rebasing.

archana-ramalingam commented on sharktank/sharktank/layers/configs/llm_configs.py:
```diff
@@ -123,7 +132,6 @@ def from_gguf_props(p: dict[str, Any]):
             attention_head_count_kv=_optional_int_prop(
                 p, f"{name_prefix}.attention.head_count_kv", attention_head_count
             ),
-            expert_count=expert_count,
```
This is necessary for all MOE models. Please add this back.

archana-ramalingam commented on sharktank/sharktank/layers/configs/llm_configs.py:
```diff
@@ -135,6 +143,22 @@ def from_gguf_props(p: dict[str, Any]):
             rope_freq_base=_optional_float_prop(
                 p, f"{name_prefix}.rope.freq_base", default_rope_freq_base
             ),
+            expert_feed_forward_length=_optional_int_prop(
```
A better way of adding architecture specific hyperparameters is to use this [function](https://github.com/nod-ai/shark-ai/blob/main/sharktank/sharktank/layers/configs/llm_configs.py#L197). 
Can be addressed in a later PR.

vivekkhandelwal1 replied:
Done.

archana-ramalingam commented on sharktank/sharktank/layers/configs/llm_configs.py:
```diff
@@ -233,6 +271,11 @@ def get_custom_configs(p: dict[str, Any], name_prefix: str):
         res["attn_head_dim"] = res["qk_nope_head_dim"] + res["qk_rope_head_dim"]
 
+    if name_prefix == "llama4":
+        res["expert_shared_count"] = _optional_int_prop(
```
This doesn't have to be `_optional_int_prop`. Because the  LlamaHParams has already set `expert_shared_count` to None by default. This holds for all params added in this `get_custom_configs()` fn.

sogartar replied:
But if we read with `_int_prop(...)` and the value is missing it will raise.

archana-ramalingam replied:
Since we have a condition that checks for llama4 architecture, it should work. 
If the model follows llama4 architecture, the model is expected to have this param.

archana-ramalingam commented on sharktank/sharktank/layers/ffn_block.py:
```diff
@@ -27,6 +27,7 @@ def __init__(
+        add_residual: bool = False,
     ):
         """
         add_residual:
```
Can remove this comment as well.

archana-ramalingam commented on sharktank/sharktank/models/llm/llm.py:
```diff
@@ -337,6 +338,11 @@ def __init__(
         fake_quant: bool = True,
     ):
+        activation_fn = (
+            ACT2FN[config.activation_fn]
```
config.activation_fn and config.ffn_add_residual are not being set in llm_configs.py for llama4 specifically. Are we addressing this in another PR or the default values align with llama4?

vivekkhandelwal1 replied:
`config.activation_fn` is not required since we support only the language version of the Llama 4 model which has a pre-determined activation function.

archana-ramalingam commented on sharktank/sharktank/models/llm/llm.py:
```diff
@@ -390,6 +396,12 @@ def __init__(
+            "llama4": (
+                torch.nn.functional.sigmoid,
+                activation_fn,
+                True,
```
This is the local equivalent of `config.ffn_add_residual`. Make sure this is the value llama4 requires. Since `config.ffn_add_residual` is not plumbed through/set in llm_configs.py, we can drop using it for now and stick to `self.add_residual` to avoid any confusion.

sogartar replied:
It is probably better to remove the `ffn_add_residual` option from the config. Hardcode it here as is done for the other models. We can add it in another PR for all model variants.

vivekkhandelwal1 commented:
> > @sogartar @archana-ramalingam I have addressed the majority of the comments. The CI is passing since I have removed 2 lines of code changes from [here](https://github.com/nod-ai/shark-ai/compare/main...users/sogartar/llama4?diff=split&w#diff-003e19a39f8d1ebd5c9f5eade128e8d3652d2526f06145730ef1c555fa9a5b17R84-R93), which was the primary cause for CI failure. I am not sure atm whether we need those changes or not for Llama 4.
> > I will test the MOE implementation with the Llama4's implementation, till then we will not merge this PR.
> 
> So, I have verified the numerics and for Llama4 Moe it works with those changes in but only for the `PreGatherFFNMOE`, not for the Dense one. Now, I need to add those changes, conditional on the Llama 4 model. But I don't see a way to identify whether the call is made for which model. Shall I add a new arg for that?
> 
> @sogartar @archana-ramalingam

I had made a change for this by adding an argument. You may review the patch now, and I can make the changes accordingly.

oyazdanb commented on sharktank/sharktank/models/llama/testing.py:
```diff
@@ -84,54 +85,66 @@ def make_attention_block_ffn_theta_v2(
+        num_experts=config.hp.expert_count,
+        with_ffn_norm=True,
+        num_shared_experts=config.hp.expert_shared_count,
+        shared_expert_hidden_dim=config.hp.expert_shared_feed_forward_length,
```
shared_expert_hidden_dim is not not defined in make_random_moe_block_theta here https://github.com/vivekkhandelwal1/shark-ai/blob/llama4/sharktank/sharktank/layers/testing.py#L314

archana-ramalingam replied:
I removed `shared_expert_hidden_dim` as it was similar to `expert_hidden_dim`. But if it's defined in the llama4 config and are diff values, we can add it back. 
PR that removed it: https://github.com/nod-ai/shark-ai/pull/1424/files#diff-01ecddc9cbe8d5f0006740065c61b914537e685b172f96b390bc529c26440d85L259

vivekkhandelwal1 replied:
I think we don't need that. If  required, I will add it in the follow-up patches.

archana-ramalingam commented:
> > > > > @sogartar @archana-ramalingam I have addressed the majority of the comments. The CI is passing since I have removed 2 lines of code changes from [here](https://github.com/nod-ai/shark-ai/compare/main...users/sogartar/llama4?diff=split&w#diff-003e19a39f8d1ebd5c9f5eade128e8d3652d2526f06145730ef1c555fa9a5b17R84-R93), which was the primary cause for CI failure. I am not sure atm whether we need those changes or not for Llama 4.
> > > > > I will test the MOE implementation with the Llama4's implementation, till then we will not merge this PR.
> > > > 
> > > > 
> > > > So, I have verified the numerics and for Llama4 Moe it works with those changes in but only for the `PreGatherFFNMOE`, not for the Dense one. Now, I need to add those changes, conditional on the Llama 4 model. But I don't see a way to identify whether the call is made for which model. Shall I add a new arg for that?
> > > > @sogartar @archana-ramalingam
> > > 
> > > 
> > > I had made a change for this by adding an argument. You may review the patch now, and I can make the changes accordingly.
> > 
> > 
> > We should pass `config.hp.model_arch` like we do for [here](https://github.com/nod-ai/shark-ai/blob/main/sharktank/sharktank/models/llm/llm.py#L362) for `PagedLlamaAttentionBlock`. This is assuming the `config.hp.model_arch==llama4`. It's sensible to add a new arg, if it's a specific op like `add_residual`, `normalize_experts` because we might expect this in future llms as well.
> 
> Hi @archana-ramalingam, I have made the changes as requested. Can you please review it now?

Great job Vivek, LGTM after resolving conflicts!
